dot_parser.py

```diff
@@ -548,8 +548,13 @@ def parse_dot_data(s):
         tokens = graphparser.parseString(s)
         return list(tokens)
     except ParseException as err:
-        print(
-            err.line +
-            " "*(err.column-1) + "^" +
-            err)
+        if PY3 and (hasattr(ParseException, 'explain') and
+                    callable(getattr(ParseException, 'explain'))):
+            print(ParseException.explain(err))
+        else:
+            print(
+                err.line + "\n" +
+                " "*(err.column-1) + "^\n" +
+                str(err)
+                )
         return None
```

Comment:
Wouldn't be better to add a new package-level config-option (e.g. `pydot.raise_errors`), not to ever return `None`,
but always let exceptions bubble-up?

Reply:
Ok, after some more reading, I certainly agree with you that `None` should not be returned.

Some sources:
- [MIT 6.0001 Introduction to Computer Science and Programming in Python, Fall 2016 by Dr. Ana Bell](https://www.youtube.com/watch?v=9H6muyZjms0&t=1678). On handling exceptions by substituting with a default value: "[T]he user thinks that they entered something, and they think everything's great, your program accepts it, but then they get some weird value as an output, which is far from what they expected. So it's not really a good idea to just replace users' values with anything."
- https://stackoverflow.com/questions/1630706/best-practice-in-python-for-return-value-on-error-vs-success
- https://github.com/williballenthin/python-registry/issues/76
- https://wiki.python.org/moin/HandlingExceptions

In our discussion in issue [#171](https://github.com/pydot/pydot/issues/171#issuecomment-612582791), I worried that some users may depend on a possible return value `None`. In other words, removing that may be a breaking change. Some remarks about that:
- I checked the current pydot 1.4.1 code and did not see anything directly or indirectly depending on a return value of `None` from this function. I also checked [NetworkX 2.4](https://github.com/networkx/networkx/tree/networkx-2.4) code as an example of a library that depends on pydot and also did not find any there. (There is of course a multitude of dependent code out there that I did not check.)
- The current docstring of the function states that it returns Graphs and does not mention possibly returning `None`.
- Still, I think changes in return values (and possible exceptions) are too far-reaching for a mere point release (e.g. 1.4.2). So, I will split off the bug fix for the `TypeError` to a separate PR (as soon as we agree on the solution for that). This PR (219) will keep the more substantial changes, including the removal of `return None`, for the next major or minor release (e.g. 1.5.x). I will add a Changelog-entry and any necessary function docstring changes to this PR in an effort to make users aware of the change.

> Wouldn't be better to add a new package-level config-option (e.g. `pydot.raise_errors`), not to ever return `None`, but always let exceptions bubble-up?

If we agree on removing `return None`, then that config-option is not necessary anymore, right? I mean, if the function does not have a Graph object to return (substitution by an empty Graph is not good either) and it should not return `None`, then what else can we do in the end except raise an exception (whether it be our own or bubbling up the original), right?

Hope this gets some things out of the way already. I will still get back on several other issues we were discussing (error handling, logging, printing, flags, etc.), either here or in our parallel discussion in issue #171.

Reply:
Thank you for the resources.
- The last one is outdated and fested with the ill-advice that troubles the newbies: if you have the code you are calling raise an exception, you should NOT catch it, unless you know what to do with it.  Besides, it's PY2.
- The 2nd from the end is a very nice discussion of a simple indexing operation: to raise or not (the inverse of the prebious).  For me, offering both possibilities, like `dict` does is the optimum.  That's simple.  But you do not ave this problem.
- The SO article confirms the pythonic consensus that state-return values are ugly.
- For the video , i dont have time, but i hope it goes further deep into the problem, and discussrs Error-Bariers (against the proliferation of unknown exception classes) or other architectural concerns and laws.

Reply:
> f we agree on removing return None, then that config-option is not necessary anymore, right? 

The suggested config-option was to retain BW-compatibilty.  If it's not needed, the it's totally fine (personally i'm ok with braking broken designs ;-)

Reply:
About that config-option `pydot.raise_errors`: We could still use it to create a transition period, with its default switching from `False` to `True` later than described above (e.g. 2.x.x). It would be similar (but inverse) to the debug/verbose flag discussed for printing in https://github.com/pydot/pydot/issues/171#issuecomment-612582791. For example, this other project started a transition period like that: https://github.com/sendgrid/sendgrid-python/issues/57.

Reply:
Sorry, I did not see your two comments before I sent out my last comment and went to sleep last night.

I will think about the BW-compatibility/transition period some more. Once we know where we want to go exactly, we might also have a better idea on how to get there. Also, I still hope to hear the opinion of the project maintainer, because he might know more about how pydot is used in practice.

> For the video , i dont have time, but i hope it goes further deep into the problem, and discussrs Error-Bariers (against the proliferation of unknown exception classes) or other architectural concerns and laws.

Not very deep, it was just an intro course. I could not find many results on Error-Bariers (or Error Barriers), do you perhaps mean [Fault Barriers](https://www.infoq.com/news/2007/01/effective-exceptions/)?

Reply:
Some more links for fault barriers:
- https://www.oracle.com/technical-resources/articles/enterprise-architecture/effective-exceptions-part1.html
- http://wiki.juneday.se/mediawiki/index.php/Design_patterns_-_Exceptions_patterns

Reply:
> Not very deep, it was just an intro course. I could not find many results on Error-Bariers (or Error Barriers), do you perhaps mean [Fault Barriers](https://www.infoq.com/news/2007/01/effective-exceptions/)?

Bullseye :+1:

Reply:
[Above](https://github.com/pydot/pydot/pull/219#discussion_r408918580), I wrote:
> I will split off the bug fix for the `TypeError` to a separate PR (as soon as we agree on the solution for that)

@ankostis: Though we did not agree on the details of it yet, I created PR #227 already, because the code of the bug fix there is no longer automatically part of the more extensive changes we are discussing here. I still welcome your comments there.